// grpc-web-client/dist/transports/xhr.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var browser_headers_1 = require("browser-headers");
var debug_1 = require("../debug");
function codePointAtPolyfill(str, index) {
    var code = str.charCodeAt(index);
    if (code >= 0xd800 && code <= 0xdbff) {
        var surr = str.charCodeAt(index + 1);
        if (surr >= 0xdc00 && surr <= 0xdfff) {
            code = 0x10000 + ((code - 0xd800) << 10) + (surr - 0xdc00);
        }
    }
    return code;
}
function stringToArrayBuffer(str) {
    var asArray = new Uint8Array(str.length);
    var arrayIndex = 0;
    for (var i = 0; i < str.length; i++) {
        var codePoint = String.prototype.codePointAt ? str.codePointAt(i) : codePointAtPolyfill(str, i);
        asArray[arrayIndex++] = codePoint & 0xFF;
    }
    return asArray;
}
exports.stringToArrayBuffer = stringToArrayBuffer;
function xhrRequest(options) {
    options.debug && debug_1.debug("xhrRequest", options);
    var xhr = new XMLHttpRequest();
    var index = 0;
    function onProgressEvent() {
        options.debug && debug_1.debug("xhrRequest.onProgressEvent.length: ", xhr.response.length);
        var rawText = xhr.response.substr(index);
        index = xhr.response.length;
        var asArrayBuffer = stringToArrayBuffer(rawText);
        options.onChunk(asArrayBuffer);
    }
    function onLoadEvent() {
        options.debug && debug_1.debug("xhrRequest.onLoadEvent");
        options.onEnd();
    }
    function onStateChange() {
        options.debug && debug_1.debug("xhrRequest.onStateChange", this.readyState);
        if (this.readyState === this.HEADERS_RECEIVED) {
            options.onHeaders(new browser_headers_1.BrowserHeaders(this.getAllResponseHeaders()), this.status);
        }
    }
    xhr.open("POST", options.url);
    xhr.responseType = "text";
    xhr.overrideMimeType("text/plain; charset=x-user-defined");
    options.headers.forEach(function (key, values) {
        xhr.setRequestHeader(key, values.join(", "));
    });
    xhr.addEventListener("readystatechange", onStateChange);
    xhr.addEventListener("progress", onProgressEvent);
    xhr.addEventListener("loadend", onLoadEvent);
    xhr.addEventListener("error", function (err) {
        options.debug && debug_1.debug("xhrRequest.error", err);
        options.onEnd(err.error);
    });
    xhr.send(options.body);
}
exports.default = xhrRequest;
